shared/src/main/scala/mlscript/NewParser.scala

```diff
@@ -859,6 +921,49 @@ abstract class NewParser(origin: Origin, tokens: Ls[Stroken -> Loc], newDefs: Bo
   
   final def exprCont(acc: Term, prec: Int, allowNewlines: Bool)(implicit et: ExpectThen, fe: FoundErr, l: Line): IfBody \/ Term = wrap(prec, s"`$acc`", allowNewlines) { l =>
     cur match {
+      case (QUOTE, l) :: _ => cur match {
+        case _ :: (KEYWORD(opStr @ "=>"), l0) :: _ if opPrec(opStr)._1 > prec =>
+          consume
+          consume
+          exprCont(Quoted(Lam(acc match {
+            case t: Tup => t
+            case _ => PlainTup(acc)
+          }, Unquoted(expr(1)(fe, implicitly)))), prec, allowNewlines)
+        case _ :: (br @ BRACKETS(Round, toks), loc) :: _ =>
+          consume
+          consume
+          val as = rec(toks, S(br.innerLoc), br.describe).concludeWith(_.argsMaybeIndented()).map {
+            case nme -> Fld(flgs, t) => nme -> Fld(flgs, Unquoted(t))
+          }
+          val res = App(Unquoted(acc), Tup(as).withLoc(S(loc)))
+          exprCont(Quoted(res), prec, allowNewlines)
+        case _ :: (IDENT(opStr, true), l0) :: _ if opPrec(opStr)._1 > prec =>
+          val currentOp = Var(opStr).withLoc(S(l0))
+          consume
+          def insQuotes(term: Term): Term = term match {
+            // * Find the correct operator and insert quotes and unquotes.
+            case App(vop @ Var(op), Tup(lhs :: rhs :: Nil)) if op === opStr && currentOp.toLoc === vop.toLoc =>
+              Quoted(App(vop, Tup((lhs :: rhs :: Nil).map {
+                case v -> Fld(flags, t) => v -> Fld(flags, Unquoted(t))
+              })))
+            // * Otherwise, search in LHS, since RHS has been done with higher priorities.
+            case App(f, Tup((v -> Fld(flags, t)) :: tail)) =>
+              App(f, Tup((v -> Fld(flags, insQuotes(t))) :: tail))
+            case Quoted(body) => Quoted(insQuotes(body))
+            case Unquoted(body) => Unquoted(insQuotes(body))
+            case _ => term
+          }
+          exprCont(acc, prec, allowNewlines) match {
+            case R(term) =>
+              R(insQuotes(term))
+            case L(bd) => bd match {
+              case IfThen(expr, rhs) =>
+                L(IfThen(insQuotes(expr), rhs))
+              case _ => L(bd)
+            }
+          }
```

Comment:
This logic is inane and is most probably incorrect in more than one way. It relies on many unstated assumptions about exactly what the recursive `exprCont` call does. The location-checking part is one of  the ugliest hacks I've seen. I expect better code quality from you by now.

Here's a better idea: in `exprCont`, use a custom `MaybeQuoted` extractor to wrap continuations that can be quoted, as in:
```diff
-  case (IDENT(opStr, true), l0) :: _ if /* isInfix(opStr) && */ opPrec(opStr)._1 > prec =>
+  case MaybeQuoted(doConsume, (IDENT(opStr, true), l0) :: _) if /* isInfix(opStr) && */ opPrec(opStr)._1 > prec =>
+    val isQuoted = doConsume() // consumes the QUOTE and returns true if there was a QUOTE
```